### ISSUE TITLE
Support carrying seclabel in IPv4 option

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#909]
- [x] Investigate feasibility of carrying seclabel as IPv4 option to not require overlay for IPv4 mode
- [x] Provide minimal BPF helper to grow headerroom at L3 (IP options)
- [x] XDP -> cls_bpf metadata transfer
- [ ] Optional: Add support to add helpers such as AH/ESP which require skb metadata change
- [ ] cilium-agent work

**Comment by bretagne-peiqi** [comment]
shall we use pure l3 network for ipv4 ?

**Comment by tgraf** [comment]
> shall we use pure l3 network for ipv4 ?

Yes, the thought behind this is to support pure L3 IPv4 mode while carrying the seclabel in the option instead of the encapsulation header. We have to investigate how feasibility it is to add/remove IPv4 options in the datapath from a performance perspective.

**Comment by borkmann** [comment]
For 4.13:

https://git.kernel.org/pub/scm/linux/kernel/git/davem/net-next.git/commit/?id=2be7e212d5419a400d051c84ca9fdd083e5aacac

Tested with clsact, ingress+egress to "decap"/"encap" 8bytes custom IP options with BPF. Node to node (nfp <-> nfp) single stream performance was about 15Gbit/s with GRO/GSO on iperf3 w/o special tuning.

**Comment by borkmann** [comment]
I get from 15 to 23Gbit/s for single stream if the receiver side does the IP option extraction in XDP, and then passes the packet up the stack. Reason is that while options work for GSO, turns out normal GRO engine has a check in fast path in `inet_gro_receive()` to stop aggregation on `if (*(u8 *)iph != 0x45)`. We could probe in cilium if the driver supports native XDP (which major 10/40G do) and then could do the sec ctx extraction and policy enforcement on XDP layer. Then we could also hit GRO fast path. For this, we would need to pass the extracted ctx meta data from XDP towards skb. iperf3 server with XDP on NIC, single stream, no special tuning:
```
[  4]   1.00-2.00   sec  2.76 GBytes  23.7 Gbits/sec    0    771 KBytes       
[  4]   2.00-3.00   sec  2.76 GBytes  23.7 Gbits/sec    0    779 KBytes       
[  4]   3.00-4.00   sec  2.76 GBytes  23.7 Gbits/sec    0    793 KBytes       
[  4]   4.00-5.00   sec  2.76 GBytes  23.7 Gbits/sec    0    806 KBytes       
[  4]   5.00-6.00   sec  2.76 GBytes  23.7 Gbits/sec    0    814 KBytes       
[  4]   6.00-7.00   sec  2.76 GBytes  23.7 Gbits/sec    0    830 KBytes       
[  4]   7.00-8.00   sec  2.76 GBytes  23.7 Gbits/sec    0    830 KBytes       
[  4]   8.00-9.00   sec  2.76 GBytes  23.7 Gbits/sec    0    830 KBytes       
[  4]   9.00-10.00  sec  2.76 GBytes  23.7 Gbits/sec    0    830 KBytes 
```

**Comment by jrfastab** [comment]
On 07/04/2017 11:09 AM, Daniel Borkmann wrote:
> I get from 15 to 23Gbit/s for single stream if the receiver side does the IP option extraction in XDP, and then passes the packet up the stack. Reason is that while options work for GSO, turns out normal GRO engine has a check in fast path in `inet_gro_receive()` to stop aggregation on `if (*(u8 *)iph != 0x45)`. We could probe in cilium if the driver supports native XDP (which major 10/40G do) and then could do the sec ctx extraction and policy enforcement on XDP layer. Then we could also hit GRO fast path. For this, we would need to pass the extracted ctx meta data from XDP towards skb. iperf3 server with XDP on NIC, single stream, no special tuning:
> ```
> [  4]   1.00-2.00   sec  2.76 GBytes  23.7 Gbits/sec    0    771 KBytes       
> [  4]   2.00-3.00   sec  2.76 GBytes  23.7 Gbits/sec    0    779 KBytes       
> [  4]   3.00-4.00   sec  2.76 GBytes  23.7 Gbits/sec    0    793 KBytes       
> [  4]   4.00-5.00   sec  2.76 GBytes  23.7 Gbits/sec    0    806 KBytes       
> [  4]   5.00-6.00   sec  2.76 GBytes  23.7 Gbits/sec    0    814 KBytes       
> [  4]   6.00-7.00   sec  2.76 GBytes  23.7 Gbits/sec    0    830 KBytes       
> [  4]   7.00-8.00   sec  2.76 GBytes  23.7 Gbits/sec    0    830 KBytes       
> [  4]   8.00-9.00   sec  2.76 GBytes  23.7 Gbits/sec    0    830 KBytes       
> [  4]   9.00-10.00  sec  2.76 GBytes  23.7 Gbits/sec    0    830 KBytes 
> ```
> 

I also need a probe for the load balancer code as well. So once that is in
place we will "know" if the NIC has support for XDP or not.

**Comment by tgraf** [comment]
This requires the new `skb_adjust_headroom` helper which will be released in 4.13. Will wait for the kernel to be released first.

**Comment by borkmann** [comment]
Full kernel support merged to land for 4.15 (#1155).
